commit 718ed6b822bf0b8f55a41b5159bdffac6f6d4b8d
Author: Slim Bouguerra <bslim@apache.org>
Date:   Thu Mar 28 15:21:56 2019 -0700

    fix the arena index start
    
    Change-Id: Ieb951ca6473189e907902d03a695a6c06b97f205

diff --git a/llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java b/llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java
index 0728f709ea..5f83a930d5 100644
--- a/llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java
+++ b/llap-server/src/java/org/apache/hadoop/hive/llap/cache/BuddyAllocator.java
@@ -281,7 +281,7 @@ public void allocateMultiple(MemoryBuffer[] dest, int size, BufferObjectFactory
     int destAllocIx = 0;
     int fastAttempt = 0;
     while (fastAttempt < MAX_DISCARD_ATTEMPTS) {
-      startArenaIx = (int)(threadId + fastAttempt  % arenaCount);
+      startArenaIx = (int)((threadId + fastAttempt) % arenaCount);
       destAllocIx =
           allocateFast(dest,
               null,
